File: Programacao_Orientada_Objetos/Abstracao/Abstracao.java
package Abstracao;
public class Abstracao {

    public static void main (String[] args) {
    
        Celular Xiaomi = new Celular();
    
        Xiaomi.ligar();
        Xiaomi.desligar();

        Detector_de_Fantasmas Super_Detector_3000 = new Detector_de_Fantasmas();

        Super_Detector_3000.ligar();
        Super_Detector_3000.detectar_fantasmas();
        Super_Detector_3000.desligar();

        Tablet TabletName = new Tablet();

        TabletName.ligar();
        TabletName.tirar_foto();
        TabletName.tirar_foto();
        TabletName.desligar();

    }
    }

abstract class Dispositivo_Eletronico {

    final String tipo;

    Dispositivo_Eletronico(String tipo) {
        this.tipo = tipo;
    };

    void ligar() {
        this.ativar_componentes_inernos();
        this.iniciar_processos();
        System.out.println(this.tipo + " ligado.");
    };

    void desligar() {
        this.desativar_componentes_inernos();
        this.fechar_processos();
        System.out.println(this.tipo + " desligado.");
    };

    private void ativar_componentes_inernos() {
        System.out.println("\nAtivando componentes do " + this.tipo);
    };

    private void desativar_componentes_inernos() {
        System.out.println("\nDesativando componentes do " + this.tipo);
    };

    private void iniciar_processos() {
        System.out.println("Iniciando processos do " + this.tipo + "\n");
    };

    private void fechar_processos() {
        System.out.println("Fechando processos do " + this.tipo + "\n");
    };
        
}

class Hand_Held extends Dispositivo_Eletronico {

    protected int bateria;
    
    Hand_Held(String tipo) {
        super(tipo);
        this.bateria = 100;
    }

    void consumir_bateria() {
        this.bateria -= 1;
    };

    void carregar_bateria() {
        this.bateria = 100;
    };

    void tirar_foto() {
        System.out.println("\nTirando foto...");
        System.out.println("Foto tirada!\n");
    }
}

class Celular extends Hand_Held {

    Celular() {
        super("Celular");
    }

}

class Tablet extends Hand_Held {

    Tablet() {
        super("Tablet");
    }
}

class Detector_de_Fantasmas extends Hand_Held {

    Detector_de_Fantasmas() {
        super("Detector de Fantasmas Falsificado");
    }

    void detectar_fantasmas() {
        System.out.println("Detectando fantasmas...");
        
        for (int x=0; x < 2; x++) {
        System.out.println("Erro! Erro! Abortar! Abortar!\nErro! Erro! Abortar! Abortar!");
        }
        System.out.println("Nenhum fAA-antassss ee-enCOOoontRAAado... (puff)");
    }

    @Override
    void tirar_foto() {
        System.out.println("Isto não era suposto a tirar fotos...");
    };
    
}

File: Programacao_Orientada_Objetos/Abstracao/abstracao.py
# A Abstração é um conceito em que você "abstrai" o código, isto é, deixa ele mais vago sobre o que exatamente
# ele faz, mas em troca, você faz com que seja mais rápido o desenvolvimento e manutenção de um código

# Suponha que temos uma função que cria uma jánela, porém existem diversas coisas que são necessárias no
# processo de criar uma jánela, como nome, processamento, renderização...
#
# Ao invés de colocar todas essas variáveis, métodos e criação de clásses espalhadas pelo código, a gente
# abstrae o código, criando uma função como:
#
# criarJanela(tamanho, nome, ...)
#
# E dentro dessa segunda função, poderiamos abstrair ainda mais as coisas, como por exemplo:
#
# processarTask(task)
# renderizarTela(janelaId)
# converterArquivo(arquivo)
#
# Com isso, a gente poderia ir até cada uma dessas funções e trabalhar nelas individualmente pra consertar
# algum erro no código delas, ao invés de ter que procurar entre diversas linhas de código espaguete


class DispositivoEletronico:
    def __init__(self, tipo):
        self.tipo = tipo

    def ligar(self):
        self._ativar_componentes_internos()
        self._iniciar_processos()
        print(f"{self.tipo} ligado.")

    def desligar(self):
        self._desativar_componentes_internos()
        self._fechar_processos()
        print(f"{self.tipo} desligado.")

    def _ativar_componentes_internos(self):
        print(f"\nAtivando componentes do {self.tipo}")

    def _desativar_componentes_internos(self):
        print(f"\nDesativando componentes do {self.tipo}")

    def _iniciar_processos(self):
        print(f"Iniciando processos do {self.tipo}\n")

    def _fechar_processos(self):
        print(f"Fechando processos do {self.tipo}\n")

class HandHeld(DispositivoEletronico):
    def __init__(self, tipo):
        super().__init__(tipo)
        self.bateria = 100

    def consumir_bateria(self):
        self.bateria -= 1

    def carregar_bateria(self):
        self.bateria = 100

    def tirar_foto(self):
        print("\nTirando foto...")
        print("Foto tirada!\n")

class Celular(HandHeld):
    def __init__(self):
        super().__init__("Celular")

class Tablet(HandHeld):
    def __init__(self):
        super().__init__("Tablet")

class DetectorDeFantasmas(HandHeld):
    def __init__(self):
        super().__init__("Detector de Fantasmas Falsificado")
        self.bb = ['a', 'b', 'c']

    def detectar_fantasmas(self):
        print("Detectando fantasmas...")
        for _ in range(2):
            print("Erro! Erro! Abortar! Abortar!\nErro! Erro! Abortar! Abortar!")
        print("Nenhum fAA-antassss ee-enCOOoontRAAado... (puff)")

    def tirar_foto(self):
        print("Isto não era suposto a tirar fotos...")

if __name__ == "__main__":
    Xiaomi = Celular()
    Xiaomi.ligar()
    Xiaomi.desligar()

    Super_Detector_3000 = DetectorDeFantasmas()
    Super_Detector_3000.ligar()
    Super_Detector_3000.detectar_fantasmas()
    Super_Detector_3000.desligar()

    Galaxy = Tablet()
    Galaxy.ligar()
    Galaxy.tirar_foto()
    Galaxy.tirar_foto()
    Galaxy.Tdesligar()
